### DISCUSSION TITLE
boolean

### DISCUSSION THREAD

**Discussion by nico-robert** [apnadkarni/tcl-cffi#136] [General]
Hello @apnadkarni ,
I have to write this function  [ImGui_ImplWin32_Init](https://github.com/ocornut/imgui/blob/9cd9c2eff99877a3f10a7f9c2a3a5b9c15ea36c6/backends/imgui_impl_win32.h#L19)
and translated to CFFI :
`DLL function ImGui_ImplWin32_Init int {hwnd pointer} ; # I believe I can write pointer.HWND too ; I have try both, same result`
As you may have seen, I replaced the `boolean` value with an `integer` value (I believe that the C language does not take this type into account).
The function does not crash, but I do not have `1` or `0` in return value, but a great value of integer e.g : `1744772353` (different value each times)
Do you have any idea ?

**Comment by apnadkarni** [comment]
The header definition that you linked seems to be a C++ header, not C. Note the type is `bool`, not `boolean`. The `bool` type in C++ is generally one byte. C99 introduced the `_Bool` type which is sometimes `#define`'ed to be `bool` in the `stdbool.h` header. You can try replacing the return type `int` with `uchar` and see if that helps.

Having said that, I think (not completely sure) the parameter passing and linkage can differ between C++ and C unless the API uses `extern "C"` or `__cdecl` in its function prototype. Calling C++ from CFFI may not be advisable and definitely not supported by libffi (and thus by cffi). Thus you should probably stick to the [C bindings](https://github.com/cimgui/cimgui) for ImGui and not the original ImGui interface.

/Ashok

**Reply by nico-robert** [reply]
Ok, I thought it was a `boolean` value, because in the .cpp file the return value was true or false like this [line](https://github.com/ocornut/imgui/blob/9cd9c2eff99877a3f10a7f9c2a3a5b9c15ea36c6/backends/imgui_impl_win32.cpp#L171)
But same in this structure [ImGuiIO](https://github.com/cimgui/cimgui/blob/068e91970db1ab543665b577a84b4f80178b1984/cimgui.h#L841) in cimgui.h by example,  there are many `bool` types. I have to replace them all with `uchar` type ?

**Reply by nico-robert** [reply]
I have tested and It works , the return value is `1` , I think I’ll create a new alias with an `enum` for that to control that I don’t have a value at `0`. `uchar` can return a value to `0` ?

**Comment by apnadkarni** [comment]
I think the right answer is to add an built-in alias to cffi for "bool" since it is compiler and platform dependent. I'll do that for the next release.

For now, you can add an alias that maps bool to uchar. At the C level (as opposed to C++), a boolean false value is 0, any other value (not just 1) is true. So you do not need a enum to control that unless I misunderstood what you were asking.

/Ashok

**Reply by apnadkarni** [reply]
Note also, `bool` is NOT the same as `Boolean` or `BOOL` that you might see on Windows. These are generally typedef'ed to `int`. Hence perhaps your initial confusion.

/Ashok

**Reply by nico-robert** [reply]
Yes, you are right , I thought than `bool` was a acronym for `boolean` value and also all `imgui.cpp` function with `bool` as return value was `true` or `false`. My bad...
